internal/application/application.go

```diff
@@ -209,6 +210,25 @@ func Delete(ctx context.Context, cluster *kubernetes.Cluster, gitea GiteaInterfa
 		return err
 	}
 
+	// get bounded services
+	app, err := Lookup(ctx, cluster, appRef.Org, appRef.Name)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	// delete all service bindings
```

Comment:
Note that we have `Bind` and `Unbind` functions for adding and removing bindings between apps and services.
See `internal/application/workload.go`.
There is even `UnbindAll` which dissolves all bindings from an app. and it is not used, there is just the definition ?!

And I thought that an `app delete` would remove the bindings already, i.e. invoke `UnbindAll` ...
But that is the reported error, right ?

I believe I will check the history for

  1. Where UnbindAll` was added, and if it was used at the time.
  2. If it was used when it was added, where that use was removed.

@jimmykarily Ping also about this.

Reply:
See 7d16af874edcd56dec66bd5ea9ba87a78147dfe0 where UnbindAll was introduced, and used.
See the following commit df9feadb287ba34417a9923986b54ff1699e728b where it got dropped.

I suspect that part of the df9fea commit has to be reverted to fix our problem.

@manno Ping.

Reply:
Relevant PR holding these commits is #553

Reply:
And we have a test for this situation: `It("unbinds bound services when deleting an app", func() {`. (See apps_test.go)

Why did that keep passing ? ... Oh, it checks only the delete output for reporting on the service, not that the binding is properly gone. And the reporting of what services are bound to the deleted app, that was not affected

Reply:
You are right. Changed code to use it.